### PR TITLE
enhancements: get list of addons model from json containing array of addons data

### PR DESCRIPTION
**Pull request by krmanik**
## Purpose / Description
In download/install addon screen, the addons are listed in with addon model info. It needs to parse addon data then addon model from json strings. 

(In next PR mapping [anki-js-addon.json](https://github.com/krmanik/anki-js-addon-json/blob/main/anki-js-addon.json) to view)
</img>

## Fixes
Fixes 

## Approach
1. Created a repository for listing and exporting all valid addons containing `ankidroid-js-addon` keywords from npmjs using GitHub actions
https://github.com/krmanik/anki-js-addon-json
2. There are two json file exported. 
    a) `anki-js-addon.json` - will be used in download/install screen, containing all addons with  `ankidroid-js-addon` keywords
    b) `test-js-addon.json` - will be used in unit test, containing only two addons info
3. Used this to map json to AddonData list 
```kotlin
val mapper = AnkiSerialization.objectMapper
val addonsData = mapper.readValue(packageJsonUrl, object : TypeReference<MutableList<AddonData>>() {})
```
4. Then used `getAddonModelFromAddonData` in loop to get all valid AddonModel and added to list
5. Finally returned list of addon model

## How Has This Been Tested?
Created unit test for this in `AddonModelTest.kt`

## Learning (optional, can help others)

_Links to blog posts, patterns, libraries or addons used to solve this problem_

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/test/java/com/ichi2/anki/jsaddons/AddonModelTest.kt [hunk 1]
```diff
@@ -115,4 +116,22 @@ class AddonModelTest : RobolectricTest() {
         assertEquals(0, reviewerEnabledAddonSet?.size)
         assertFalse(reviewerEnabledAddonSet!!.contains(addonModel.name))
     }
+
+    @Test
+    fun getAddonModelListFromJsonTest() {
+        shadowOf(getMainLooper()).idle()
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/anki/jsaddons/AddonModelTest.kt:122]: Shouldn't be needed

File: AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonData.kt [hunk 2]
```diff
@@ -135,3 +148,30 @@ fun getAddonModelFromJson(packageJsonPath: String): Pair<AddonModel?, List<Strin
...
+
+        if (result.first == null) {
+            Timber.i("Not a valid addon for AnkiDroid, the errors for the addon:\n %s", result.second.toString())
+            errorList.add(result.second.toString())
+            continue
+        }
+        addonsModelList.add(result.first!!)
+    }
+
+    return Pair(addonsModelList, ImmutableList(errorList))
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonData.kt:176]:
```suggestion
    return Pair(addonsModelList, errorList.toList())
```

File: AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonData.kt [hunk 3]
```diff
@@ -135,3 +148,30 @@ fun getAddonModelFromJson(packageJsonPath: String): Pair<AddonModel?, List<Strin
...
+    val mapper = AnkiSerialization.objectMapper
+    val addonsData = mapper.readValue(packageJsonUrl, object : TypeReference<MutableList<AddonData>>() {})
+
+    val addonsModelList = mutableListOf<AddonModel>()
+    for (addon in addonsData) {
+        val result = getAddonModelFromAddonData(addon)
+
+        if (result.first == null) {
+            Timber.i("Not a valid addon for AnkiDroid, the errors for the addon:\n %s", result.second.toString())
+            errorList.add(result.second.toString())
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonData.kt:170]: interesting: one "string" per addon, maybe we want an `errorList.addAll()` here 

Implementer/Reviewer2's choice

File: AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonData.kt [hunk 4]
```diff
@@ -135,3 +148,30 @@ fun getAddonModelFromJson(packageJsonPath: String): Pair<AddonModel?, List<Strin
...
+ * Get list of AddonModel from json containing arrays of addons package info
+ *
+ * @param packageJsonUrl package json url containing arrays of addon packages info
+ * @return Pair with list of valid addonModel and error list
+ */
+@Throws(IOException::class)
+fun getAddonModelListFromJson(packageJsonUrl: URL): Pair<List<AddonModel>, List<String>> {
+    val errorList: MutableList<String> = ArrayList()
+    val mapper = AnkiSerialization.objectMapper
+    val addonsData = mapper.readValue(packageJsonUrl, object : TypeReference<MutableList<AddonData>>() {})
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonData.kt:162]: can you comment above that this is a network request

File: AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonData.kt [hunk 5]
```diff
@@ -135,3 +148,30 @@ fun getAddonModelFromJson(packageJsonPath: String): Pair<AddonModel?, List<Strin
...
+    val addonsModelList = mutableListOf<AddonModel>()
+    for (addon in addonsData) {
+        val result = getAddonModelFromAddonData(addon)
+
+        if (result.first == null) {
+            Timber.i("Not a valid addon for AnkiDroid, the errors for the addon:\n %s", result.second.toString())
+            errorList.add(result.second.toString())
+            continue
+        }
+        addonsModelList.add(result.first!!)
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonData.kt:173]: Hmm... I'm surprised the !! is needed. There may be a bug, but it looks fine

File: AnkiDroid/src/test/java/com/ichi2/anki/jsaddons/AddonModelTest.kt [hunk 6]
```diff
@@ -115,4 +116,22 @@ class AddonModelTest : RobolectricTest() {
...
+
+        val url = URL("https://raw.githubusercontent.com/krmanik/anki-js-addon-json/main/test-js-addon.json")
+        val result = getAddonModelListFromJson(url)
+
+        // first addon name and tgz download url
+        val addon1 = result.first[0]
+        assertEquals(addon1.name, "ankidroid-js-addon-progress-bar")
+        addon1.dist["tarball"]?.let { assertTrue(it.endsWith(".tgz")) }
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/anki/jsaddons/AddonModelTest.kt:130]: Shouldn't this test fail if the tarball is null?

**Review by david-allison — CHANGES_REQUESTED** [review]
small nitpicks, seems reasonable in general

File: AnkiDroid/src/test/java/com/ichi2/anki/jsaddons/AddonModelTest.kt (hunk 6 above)
Comment (krmanik) [reply AnkiDroid/src/test/java/com/ichi2/anki/jsaddons/AddonModelTest.kt]: Change the test null pointer exception will thrown when null

Comment (krmanik) [reply AnkiDroid/src/test/java/com/ichi2/anki/jsaddons/AddonModelTest.kt]: Change the test null pointer exception will thrown when null

File: AnkiDroid/src/test/java/com/ichi2/anki/jsaddons/AddonModelTest.kt [hunk 7]
```diff
@@ -119,19 +119,17 @@ class AddonModelTest : RobolectricTest() {
...
-        shadowOf(getMainLooper()).idle()
-
         val url = URL("https://raw.githubusercontent.com/krmanik/anki-js-addon-json/main/test-js-addon.json")
         val result = getAddonModelListFromJson(url)
 
         // first addon name and tgz download url
         val addon1 = result.first[0]
         assertEquals(addon1.name, "ankidroid-js-addon-progress-bar")
-        addon1.dist["tarball"]?.let { assertTrue(it.endsWith(".tgz")) }
+        assertTrue(addon1.dist["tarball"]!!.endsWith(".tgz"))
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/anki/jsaddons/AddonModelTest.kt:128]: `assertThat(, endsWith())` - gives a better message

Comment (krmanik) [reply AnkiDroid/src/test/java/com/ichi2/anki/jsaddons/AddonModelTest.kt]: Done, Thanks

File: AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonData.kt [hunk 8]
```diff
@@ -135,3 +147,31 @@ fun getAddonModelFromJson(packageJsonPath: String): Pair<AddonModel?, List<Strin
...
+    val mapper = AnkiSerialization.objectMapper
+    // network request to fetch json from given url
+    val addonsData = mapper.readValue(packageJsonUrl, object : TypeReference<MutableList<AddonData>>() {})
+
+    val addonsModelList = mutableListOf<AddonModel>()
+    for (addon in addonsData) {
+        val result = getAddonModelFromAddonData(addon)
+
+        if (result.first == null) {
+            Timber.i("Not a valid addon for AnkiDroid, the errors for the addon:\n %s", result.second.toString())
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonData.kt:169]:
```suggestion
            Timber.i("Not a valid addon for AnkiDroid, the errors for the addon:\n %s", result.second)
```

File: AnkiDroid/src/test/java/com/ichi2/anki/jsaddons/AddonModelTest.kt [hunk 9]
```diff
@@ -115,4 +118,20 @@ class AddonModelTest : RobolectricTest() {
         assertEquals(0, reviewerEnabledAddonSet?.size)
         assertFalse(reviewerEnabledAddonSet!!.contains(addonModel.name))
     }
+
+    @Test
+    fun getAddonModelListFromJsonTest() {
+        val url = URL("https://raw.githubusercontent.com/krmanik/anki-js-addon-json/main/test-js-addon.json")
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/anki/jsaddons/AddonModelTest.kt:124]: Second reviewer: flagging that this unit test needs network access

**Review by david-allison — APPROVED** [review]

File: AnkiDroid/src/test/java/com/ichi2/anki/jsaddons/AddonModelTest.kt (hunk 9 above)
Comment (krmanik) [reply AnkiDroid/src/test/java/com/ichi2/anki/jsaddons/AddonModelTest.kt:124]: I have put json and tgz file in assets folder. I will also move this to assets folder.

Comment (krmanik) [reply AnkiDroid/src/test/java/com/ichi2/anki/jsaddons/AddonModelTest.kt]: Moved to assets folder.

File: AnkiDroid/src/test/java/com/ichi2/anki/jsaddons/AddonModelTest.kt [hunk 10]
```diff
@@ -54,6 +58,7 @@ class AddonModelTest : RobolectricTest() {
 
         validNpmPackageJson = FileOperation.getFileResource("valid-ankidroid-js-addon-test.json")
         notValidNpmPackageJson = FileOperation.getFileResource("not-valid-ankidroid-js-addon-test.json")
+        addonsPackageListTestJson = FileOperation.getFileResource("test-js-addon.json")
```
Comment (mikehardy) [inline AnkiDroid/src/test/java/com/ichi2/anki/jsaddons/AddonModelTest.kt:61]: test does not need network access, thankfully @david-allison

File: AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonData.kt (hunk 4 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonData.kt:162]: The test itself has a JSON in test resources, which is nice as it won't have external resources required
And then this method takes the JSON URL as a param which is also nice. I'm not against having a curated collection of them like in the krmanik repo but I would like to have the URL be configurable or for there to be an easy way to do a wide search on npmjs.com or whatever URL people want. The idea being to make the easy stuff easy (krmanik repo) but the hard stuff possible (install some random package from whatever new package repo in the future)

**Review by mikehardy — APPROVED** [review]
This looks good to me - all my comments were just adding context as I understand it

**Comment by mikehardy** [comment]
For anyone watching order of operations here - with all the kotlin migrations going in - which have a tendency to create cross-PR build failures on main when merged, I'm pulling every PR that goes "pending merge", running all tests from CI locally, and then merging them when they pass, one by one. So I am occasionally merging stuff before GitHub Actions is passing but I'm definitely running tests....

**Comment by krmanik** [comment]
Thanks

**Pull request merged**